.travis.yml

```diff
@@ -65,6 +73,11 @@ jobs:
     - name: "Formatting & clippy"
       # https://github.com/travis-ci/travis-ci/issues/5175#issuecomment-215016943
       language: bash
+
+      # sudo is needed for codecov, && system library install which makes things 20-50s slower
+      # https://docs.travis-ci.com/user/reference/overview/#virtualisation-environment-vs-operating-system
+      sudo: true
```

Comment:
@neonphog do we absolutely need sudo? as the comment says, this adds 20-50s to each build you put it on

this extra time doesn't show up in the build logs, it's "boot time"

Reply:
@thedavidmeister  – I have to say I’m getting pretty frustrated with this process. I super appreciate your dedication to having a solid CI pipeline, we absolutely need something that is quick and accurate. At the same time, though, I’ve now spent more work hours going back and forth with Travis than I did writing the original code. We aren’t going to be able to reach our goals if we continue at this pace. Any thoughts or suggestions you have would be much appreciated!

To install the libzmq3-dev system library we need `sudo`. * Somehow * when that command was in the  `.travis.yml` file, it worked without the sudo required. I don’t know if travis has a special exception for `before_script` that allows sudo in there, or if the coverage task was somehow caching the library for the other tasks, or what… In any case, when I moved the install into the Makefile, the three jobs that did not have sudo started failing.

Reply:
@neonphog yes, we've been moving things through more quickly as smaller PRs

is it possible for this to be smaller than 1k LOC?

i spent quite a few hours speeding travis up

Reply:
@thedavidmeister - are you asking me to break this up into multiple pull requests?

github is saying `+1,018 −7`, the Makefile changes alone are `35` LOC... so I could split out the stuff for getting `libzmq3-dev` installed and that would bring the remainder below 1000.

Do you have any suggestions about the sudo? I could make my own binary build for Ubuntu Trusty in a separate repo, link it to holochain-rust as a git submodule and only in the TRAVIS case set up environment variables that cause the zmq rust crate to link against that submodule binary... that would not require sudo, though I'm not sure how reliable it would be.

Reply:
Yeah, on second thought, i cannot do that. It's actually 5 libs we'd need to maintain, that's not the business we want to be in....

```
Get:2 http://download.opensuse.org/repositories/network:/messaging:/zeromq:/release-stable/xUbuntu_14.04 ./ libnorm1 1.5r6+dfsg1-9 [219 kB]
Get:3 http://download.opensuse.org/repositories/network:/messaging:/zeromq:/release-stable/xUbuntu_14.04 ./ libpgm-5.2-0 5.2.122~dfsg-2 [157 kB]
Get:4 http://download.opensuse.org/repositories/network:/messaging:/zeromq:/release-stable/xUbuntu_14.04 ./ libsodium18 1.0.11-1 [154 kB]
Get:5 http://download.opensuse.org/repositories/network:/messaging:/zeromq:/release-stable/xUbuntu_14.04 ./ libzmq5 4.2.5 [217 kB]
Get:6 http://download.opensuse.org/repositories/network:/messaging:/zeromq:/release-stable/xUbuntu_14.04 ./ libzmq3-dev 4.2.5 [398 kB]
```

Reply:
aha, @thedavidmeister - according to this https://github.com/travis-ci/travis-ci/issues/8877 - if travis detects a sudo command in the .travis.yml file, then sudo is auto-enabled... that explains why it was working in the before_script at least... it * was * actually enabling sudo for all the tasks

Reply:
So you need `sudo` for the `sudo apt-get install -y libzmq3-dev;` down there in Makefile, right? I think that should go into our image anyway: https://github.com/holochain/holochain-rust/blob/develop/docker/Dockerfile.ubuntu

right? @neonphog @thedavidmeister

Reply:
Docker is no longer used in travic-ci @lucksus - I * did * add that line to the docker image. What makes it worse is that Travis is still using a really old version of Ubuntu, so we cannot just apt-get the system lib (the version is incompatible with the rust crate) it's the four TRAVIS lines to install the binary packages built by the libzmq folks